tracer/src/Datadog.Trace/AppSec/WafEncoding/Encoder.cs

```diff
@@ -63,442 +63,459 @@ public static string FormatArgs(object o)
-                            var itemData = childrenData;
-                            var idx = 0;
-                            foreach (var val in enumerable)
-                            {
-                                if (idx > childrenCount)
-                                {
-                                    break;
-                                }
-
-                                *(DdwafObjectStruct*)itemData = Encode(val, argToFree, applySafetyLimits: applySafetyLimits, remainingDepth: remainingDepth, pool: pool);
-                                itemData += ObjectStructSize;
-                                idx++;
-                            }
-
-                            ddwafObjectStruct.Array = childrenData;
-                            ddwafObjectStruct.NbEntries = (ulong)childrenCount;
-                            argToFree.Add(childrenData);
+                            sb.Remove(sb.Length - 2, 2);
                         }
+
+                        return StringBuilderCache.GetStringAndRelease(sb);
                     }
 
-                    break;
+                    Log.Warning("EncodeDictionary: object graph too deep, truncating nesting {Items}", GetItemsAsString());
+                    return ddWafObjectMap;
                 }
 
-                default:
-                    if (Log.IsEnabled(Vendors.Serilog.Events.LogEventLevel.Debug))
+                if (count > WafConstants.MaxContainerSize)
+                {
+                    Log.Warning<int>("EncodeList: list too long, it will be truncated, MaxMapOrArrayLength {MaxMapOrArrayLength}", WafConstants.MaxContainerSize);
+                }
+            }
+
+            var childrenCount = !context.ApplySafetyLimits || count < WafConstants.MaxContainerSize ? count : WafConstants.MaxContainerSize;
+            var childrenFromPool = ObjectStructSize * childrenCount < MaxBytesForMaxStringLength;
+            var childrenData = childrenFromPool ? context.Pool.Rent() : Marshal.AllocCoTaskMem(ObjectStructSize * childrenCount);
+
+            if (enumerableDic is IDictionary iDic)
+            {
+                var typeKVP = typeof(KeyValuePair<TKey, TValue>);
+                if (typeKVP == typeof(KeyValuePair<string, string>))
+                {
+                    EnumerateIDictionaryItems<string, string>(
+                        ref context,
+                        remainingDepth,
+                        iDic,
+                        (delegate*< KeyValuePair<string, string>, string?>)getKey,
+                        (delegate*<KeyValuePair<string, string>, object?>)getValue,
+                        childrenData,
+                        childrenCount);
+                }
+                else if (typeKVP == typeof(KeyValuePair<string, object>))
+                {
+                    EnumerateIDictionaryItems<string, object>(
```

Comment:
NIT: redundant type arguments
```suggestion
                    EnumerateIDictionaryItems(
```